### ISSUE TITLE
Type alias used in async fn return type is wrongly reported unused

### ISSUE THREAD

**Issue by Globidev** [rust-lang/rust#59085]
With `rustc 1.35.0-nightly (26b4cb484 2019-03-09)`
The following code
```rust
type Bar = ();
async fn foo() -> Bar { }
```
produces the following warning
```
warning: type alias is never used: `Bar`                                  
 --> src/main.rs:3:1
  |
3 | type Bar = ();
  | ^^^^^^^^^^^^^^
  |
  = note: #[warn(dead_code)] on by default
```
When `Bar` is clearly used here.
This seem to only trigger if the alias is used in the return type position

**Comment by sanxiyn** [comment]
This is unrelated to async fn as one can reproduce with just impl Trait:

```rust
trait Trait {
    type Type;
}

impl Trait for () {
    type Type = ();
}

type Bar = ();
fn foo() -> impl Trait<Type = Bar> {}

fn main() {
    foo();
}
```
